orderbook/src/maintenance.rs

```diff
@@ -0,0 +1,34 @@
+use crate::database::Database;
+use crate::event_updater::EventUpdater;
+use crate::orderbook::Orderbook;
+use anyhow::Result;
+use futures::try_join;
+use shared::current_block::Maintaining;
+use std::sync::Arc;
+
+pub struct ServiceMaintenance {
+    storage: Arc<Orderbook>,
+    database: Database,
+    event_updater: EventUpdater,
+}
+
+impl ServiceMaintenance {
+    pub fn new(storage: Arc<Orderbook>, database: Database, event_updater: EventUpdater) -> Self {
+        ServiceMaintenance {
+            storage,
+            database,
+            event_updater,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Maintaining for ServiceMaintenance {
+    async fn run_maintenance(&self) -> Result<()> {
+        try_join!(
+            self.storage.run_maintenance(),
+            self.event_updater.run_maintenance()
+        )?;
+        self.database.run_maintenance().await
```

Comment:
Just for the other reviewers, we intentionally only make the storage and event updater run concurrently followed serially by the database maintenance as that was how the original `orderbook_maintenance` was run. I don't know if its actually necessary to run this way, or if they all can run concurrently.

Note that both the `event_updater` and the `database` maintenance can cause changes to the Prostgres database.

Reply:
I could add a comment that this is why we have kept them separate here.

Reply:
It might also be good to know if its still necessary.

Reply:
While it is probably safer to keep these from being run in parallel, the event updater primarily targets (but it not limited to) writes to the trade table, it also updates existing orders (on `OrderInvalidation`). database maintenance is mostly concerned with removing expired fee quotes.

It is possible that future iterations could change either of these places and a project maintainer may not think to look here, so its probably safer to just keep them apart with a comment.

Reply:
They can run at the same time but also it doesn't really matter if they for maintenance that happens every 10 seconds. So I would remove the comment saying that they shouldn't but also it's fine to keep them sequential here.

Reply:
One disadvantage of try_join is that

> try_join! is similar to join!, but completes immediately if any of the futures return an error.

So if one of them fails it will cancel the others. Imo it would be better if they all get a chance to run. This is what the previous code did. It just logged the errors but ran all of them.

Reply:
Cool, so I will remove the comment and use `join!` instead.

Reply:
I have one question here about what might be best for this.

We need to resolve a Tuple `(Result<(), anyhow::Error>, Result<(), anyhow::Error>)`
because of unused results. In the previous implementation we had just gone with:

```rust
        join!(
            self.storage.run_maintenance(),
            self.event_updater.run_maintenance()
        ).0?;
```

but this is giving preference to the first result and seems like cheating... should I assign this like `res = join!(one, two)` and then do 
```
res.0?;
res.1?;
``` 
in order to be unbiased?

Reply:
In that case, @vkgnosis do you think we should just `join` on all 3 futures, and then print out any of the errors that we encountered?. Something like:
```rust
for result of future::join_all(vec![
            self.storage.run_maintenance(),
            self.event_updater.run_maintenance(),
            self.database.run_maintenance(),
]) {
  result?;
}
```

Reply:
Right... I just had to make it a vector and use join_all!

Seems good!

Reply:
> but this is giving preference to the first result and seems like cheating...

Remember that this used to do `balance_fetcher.update().await` which can't fail (see `impl Maintaining for Orderbook`) which is why it worked like that before. I think we can ether:

```rust
let result = join!(
    self.storage.run_maintenance(),
    self.event_updater.run_maintenance(),
    self.database.run_maintenance(),
);
result.0?;
result.1?;
result.2?;
```

Or my suggestion above.

Reply:
I would say each individual task should be run and if it errors, the error logged. Then you could always return Ok from the bigger function.
Something like join_all should be fine.